File: content/17-math-and-geometry/spiral_matrix.js
/**
 * Problem: Spiral Traversal on a Matrix
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/spirally-traversing-a-matrix/0
 */

function spiralOrder(matrix) {
    /**
     * Time Complexity: O(M * N)
     * Space Complexity: O(1) auxiliary
     */
    if (!matrix || matrix.length === 0 || matrix[0].length === 0) return [];
    
    const res = [];
    let top = 0, bottom = matrix.length - 1;
    let left = 0, right = matrix[0].length - 1;
    
    while (top <= bottom && left <= right) {
        // Traverse right
        for (let col = left; col <= right; col++) {
            res.push(matrix[top][col]);
        }
        top++;
        
        // Traverse down
        for (let row = top; row <= bottom; row++) {
            res.push(matrix[row][right]);
        }
        right--;
        
        // Traverse left
        if (top <= bottom) {
            for (let col = right; col >= left; col--) {
                res.push(matrix[bottom][col]);
            }
            bottom--;
        }
        
        // Traverse up
        if (left <= right) {
            for (let row = bottom; row >= top; row--) {
                res.push(matrix[row][left]);
            }
            left++;
        }
    }
    
    return res;
}

if (require.main === module) {
    const assert = require('assert');
    const mat = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]];
    assert.deepStrictEqual(spiralOrder(mat), [1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7]);
    console.log("JS: spiralOrder passed tests!");
}

File: content/17-math-and-geometry/spiral_matrix.py
"""
Problem: Spiral Traversal on a Matrix
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/spirally-traversing-a-matrix/0
"""

def spiral_order(matrix: list[list[int]]) -> list[int]:
    """
    Time Complexity: O(M * N)
    Space Complexity: O(1) auxiliary
    """
    if not matrix or not matrix[0]:
        return []
        
    res = []
    top, bottom = 0, len(matrix) - 1
    left, right = 0, len(matrix[0]) - 1
    
    while top <= bottom and left <= right:
        # Traverse right
        for col in range(left, right + 1):
            res.append(matrix[top][col])
        top += 1
        
        # Traverse down
        for row in range(top, bottom + 1):
            res.append(matrix[row][right])
        right -= 1
        
        # Traverse left
        if top <= bottom:
            for col in range(right, left - 1, -1):
                res.append(matrix[bottom][col])
            bottom -= 1
            
        # Traverse up
        if left <= right:
            for row in range(bottom, top - 1, -1):
                res.append(matrix[row][left])
            left += 1
            
    return res

if __name__ == "__main__":
    mat = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
    assert spiral_order(mat) == [1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7]
    print("Python: spiral_order passed tests!")
